sdk/eventhubs/azure-messaging-eventhubs/src/main/java/com/azure/messaging/eventhubs/EventHubBufferedPartitionProducer.java

```diff
@@ -226,83 +226,114 @@ private EventDataBatch createNewBatch() {
+            Consumer<SendBatchSucceededContext> onSucceed, Consumer<SendBatchFailedContext> onFailed,
             Duration operationTimeout, ClientLogger logger) {
             this.partitionId = partitionId;
+            this.eventSink = eventSink;
             this.onSucceed = onSucceed;
             this.onFailed = onFailed;
-            this.flushSemaphore = flushSemaphore;
-            this.isFlushing = flush;
             this.operationTimeout = operationTimeout;
             this.logger = logger;
         }
 
         @Override
-        protected void hookOnSubscribe(Subscription subscription) {
+        protected void hookOnSubscribe(Subscription s) {
             requestUnbounded();
         }
 
         @Override
         protected void hookOnNext(PublishResult result) {
-            if (result.error == null) {
-                onSucceed.accept(new SendBatchSucceededContext(result.batch.getEvents(), partitionId));
-            } else {
-                onFailed.accept(new SendBatchFailedContext(result.batch.getEvents(), partitionId, result.error));
+            if (result != PublishResult.EMPTY) {
+                if (result.error == null) {
+                    onSucceed.accept(new SendBatchSucceededContext(result.events, partitionId));
+                } else {
+                    onFailed.accept(new SendBatchFailedContext(result.events, partitionId, result.error));
+                }
             }
-
-            tryCompleteFlush();
+            final FlushCompletionOrigin origin
+                = (result == PublishResult.EMPTY) ? FlushCompletionOrigin.ON_NEXT_EMPTY : FlushCompletionOrigin.ON_NEXT;
+            tryCompleteFlush(origin);
         }
 
         @Override
-        protected void hookOnError(Throwable throwable) {
-            logger.atError()
-                .addKeyValue(PARTITION_ID_KEY, partitionId)
-                .log("Publishing subscription completed and ended in an error.", throwable);
-
-            onFailed.accept(new SendBatchFailedContext(null, partitionId, throwable));
-
-            tryCompleteFlush();
+        protected void hookOnError(Throwable t) {
+            if (terminated.getAndSet(true)) {
+                return;
+            }
+            logger.atError().log("Publishing-subscription terminated with an error.", t);
+            onFailed.accept(new SendBatchFailedContext(null, partitionId, t));
+            tryCompleteFlush(FlushCompletionOrigin.TERMINAL_ERROR);
         }
 
         @Override
         protected void hookOnComplete() {
-            logger.atInfo().addKeyValue(PARTITION_ID_KEY, partitionId).log("Publishing subscription completed.");
+            if (terminated.getAndSet(true)) {
+                return;
+            }
+            logger.atInfo().log("Publishing-subscription terminated.");
+            tryCompleteFlush(FlushCompletionOrigin.TERMINAL_COMPLETION);
+        }
 
-            tryCompleteFlush();
+        TimeoutException awaitPendingFlush() {
```

Comment:
This is the original code existed in `enqueueEvent` but as a function in `PublishResultSubscriber`, to reduce the scope of semaphore and atomic for the flush local to this class.